import_obfuscation.go

```diff
@@ -0,0 +1,572 @@
+	namedata
+)
+
+// privateImports stores package paths and names that
+// match GOPRIVATE. privateNames are elements of the
+// paths in privatePaths, separated so that the shorter
+// names don't accidently match another import, such
+// as a stdlib package
+type privateImports struct {
+	privatePaths []string
+	privateNames []string
+}
+
+func obfuscateImports(objPath, importCfgPath string) (map[string]string, error) {
+	importCfg, err := goobj2.ParseImportCfg(importCfgPath)
+	if err != nil {
+		return nil, err
+	}
+	mainPkg, err := goobj2.Parse(objPath, "main", importCfg)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing main objfile: %v", err)
+	}
+	privatePkgs := []pkgInfo{{mainPkg, objPath}}
+
+	// build list of imported packages that are private
+	for pkgPath, info := range importCfg {
+		if isPrivate(pkgPath) {
+			pkg, err := goobj2.Parse(info.Path, pkgPath, importCfg)
+			if err != nil {
+				return nil, fmt.Errorf("error parsing objfile %s at %s: %v", pkgPath, info.Path, err)
+			}
+
+			privatePkgs = append(privatePkgs, pkgInfo{pkg, info.Path})
+		}
+	}
+
+	var sb strings.Builder
+	var buf bytes.Buffer
+	garbledImports := make(map[string]string)
+	for _, p := range privatePkgs {
+		// log.Printf("++ Obfuscating object file for %s ++", p.pkg.ImportPath)
+		for _, am := range p.pkg.ArchiveMembers {
+			// log.Printf("\t## Obfuscating archive member %s ##", am.ArchiveHeader.Name)
+
+			// skip objects that are not used by the linker, or that do not contain
+			// any Go symbol info
+			if am.IsCompilerObj() || am.IsDataObj() {
+				continue
+			}
+
+			// add all private import paths to a list to garble
+			var privImports privateImports
+			privImports.privatePaths, privImports.privateNames = explodeImportPath(p.pkg.ImportPath)
+			// the main package might not have the import path "main" due to modules,
+			// so add "main" to private import paths
+			if p.pkg.ImportPath == buildInfo.firstImport {
+				privImports.privatePaths = append(privImports.privatePaths, "main")
+			}
+
+			initImport := func(imp string) string {
+				if !isPrivate(imp) {
+					return imp
+				}
+
+				privPaths, privNames := explodeImportPath(imp)
+				privImports.privatePaths = append(privImports.privatePaths, privPaths...)
+				privImports.privateNames = append(privImports.privateNames, privNames...)
+
+				return hashImport(imp, garbledImports)
+			}
+
+			for i := range am.Imports {
+				am.Imports[i].Pkg = initImport(am.Imports[i].Pkg)
+			}
+			for i := range am.Packages {
+				am.Packages[i] = initImport(am.Packages[i])
+			}
+
+			// move imports that contain another import as a substring to the front,
+			// so that the shorter import will not match first and leak part of an
+			// import path
+			sort.Slice(privImports.privatePaths, func(i, j int) bool {
+				iSlashes := strings.Count(privImports.privatePaths[i], "/")
+				jSlashes := strings.Count(privImports.privatePaths[j], "/")
+				// sort by number of slashes first, then alphabetically
```

Comment:
This comment doesn't quite fit, because the alphabetic sort appears before the slash sort. Either rework the comment or move the slash sort before the alphabetic sort.